engine/battle/effect_commands.asm

```diff
@@ -4823,9 +4823,43 @@ UpdateMoveData:
 	push af
 
 	; Write to move selection
+	; Support for duplicate moves: If wCur[Enemy]MoveNum already points to the
+	; current move, preserve it. This allows trainers to have multiple instances
+	; of the same move (e.g., Lance's Dragonite with two Hyper Beams).
 	ld c, a
 	ld hl, wBattleMonMoves
 	call GetUserMonAttr
+
+	; Check if the current move number already points to the correct move
+	push hl
+	ldh a, [hBattleTurn]
+	and a
+	ld hl, wCurMoveNum
+	jr z, .check_current_move_num
+	ld hl, wCurEnemyMoveNum
+.check_current_move_num
+	ld a, [hl]
```

Comment:
This could save the `push hl/pop hl`. Given that all four registers were already pushed to the stack in this whole segment, I find even more pushes/pops to be suspicious (surely it has enough free registers for this bit of work).
```suggestion
	ld a, [wCurMoveNum]
	jr z, .check_current_move_num
	ld a, [wCurEnemyMoveNum]
.check_current_move_num
```